src/hotspot/share/gc/g1/g1ConcurrentRebuildAndScrub.cpp

```diff
@@ -87,14 +87,19 @@ class G1RebuildRSAndScrubTask : public WorkerTask {
       _processed_words += processed;
     }
 
-    // Yield if enough has been processed; returns if the concurrent marking cycle
-    // has been aborted for any reason.
-    bool yield_if_necessary() {
+    // Yield if enough has been processed. Return whether we should stop
+    // processing this region because either the concurrent marking cycle has been
+    // aborted or the region has been reclaimed.
+    bool yield_if_necessary(G1HeapRegion* hr) {
       if (_processed_words >= ProcessingYieldLimitInWords) {
         reset_processed_words();
-        _cm->do_yield_check();
+        // If a yield occurs (potential young-gc pause), must recheck for
+        // potential regions reclamation.
+        if (_cm->do_yield_check() && !should_rebuild_or_scrub(hr)) {
+          return true;
+        }
       }
-      return _cm->has_aborted();
+      return _cm->has_aborted() || !should_rebuild_or_scrub(hr);
```

Comment:
If I don't miss anything, this is the difference between the previous patch and this one.

Reply:
Yes, this is what the second commit in this PR contains.